File: packages/canvas/src-native/canvas-android/canvas/src/main/java/com/github/triniwiz/canvas/AnimationFrame.java
package com.github.triniwiz.canvas;

import android.animation.TimeAnimator;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;
import android.os.SystemClock;
import android.util.Log;
import android.view.Choreographer;

import java.nio.LongBuffer;
import java.sql.Array;
import java.time.Instant;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executor;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

/**
 * Created by triniwiz on 2019-08-13
 */
public class AnimationFrame implements Choreographer.FrameCallback {
    private static HashMap<Long, Callback> callbacks;
    static long lastCall = 0;
    static int count = 0;
    static Timer timer;
    static AnimationFrame instance;
    static ExecutorService executorService = Executors.newCachedThreadPool();
    static final Object lock = new Object();
    static Choreographer choreographer;
    static Handler handler;
    static long animationId = 0;

    static {
        callbacks = new HashMap<>();
        instance = new AnimationFrame();
        handler = new Handler(Looper.getMainLooper());
    }

    static int _minFps = 1000;
    static int framesRendered = 0;
    static long frameStartTime = 0;
    static boolean inAnimationFrame = false;

    void reset() {
        _minFps = 1000;
        frameStartTime = 0;
        framesRendered = 0;
    }

    void raf(final long fps){
        executorService.submit(new Runnable() {
            @Override
            public void run() {
                HashMap<Long, Callback> cbs = (HashMap<Long, Callback>) callbacks.clone();
                callbacks.clear();
                inAnimationFrame = true;
                Set<Map.Entry<Long, Callback>> set = cbs.entrySet();
                for (final Map.Entry<Long, Callback> cb : set) {
                    //if(cb.getKey() == animationId){
                    handler.post(new Runnable() {
                        @Override
                        public void run() {
                            cb.getValue().onFrame(fps);
                        }
                    });
                    // }
                }
                inAnimationFrame = false;
                reset();
            }
        });
    }

    void fps(long currentTimeMillis) {
        int fps;
        if (frameStartTime > 0) {
            // take the span in milliseconds
            long timeSpan = (currentTimeMillis - frameStartTime);
            framesRendered++;

            if (timeSpan > 1000) {
                fps = (int) (framesRendered * 1000 / timeSpan);
                if (fps < _minFps) {
                    _minFps = fps;
                }
                raf(fps);
                frameStartTime = currentTimeMillis;
                framesRendered = 0;
            }
        } else {
            frameStartTime = currentTimeMillis;
        }

    }


    @Override
    public void doFrame(long frameTimeNanos) {
        fps(frameTimeNanos / 1000000);
        Choreographer.getInstance().postFrameCallback(instance);
    }

    public interface Callback {
        void onFrame(long called);
    }

    private static long getTimeInFrameBase() {
        return System.nanoTime() / 1000000;
    }

    private static long getNewId() {
        animationId++;
        return animationId;
    }

    public synchronized static long requestAnimationFrame(Callback callback) {
        if (!inAnimationFrame) {
            inAnimationFrame = true;
            callback.onFrame(getTimeInFrameBase());
            inAnimationFrame = false;
            return getNewId();
        }
        long id = getNewId();
        callbacks.put(id, callback);
        Choreographer.getInstance().postFrameCallback(instance);
        return id;
    }


    public static void cancelAnimationFrame(long id) {
        callbacks.remove(id);
    }
}

File: packages/canvas/src-native/canvas-android/canvas/src/main/java/org/nativescript/canvas/AnimationFrame.kt
package org.nativescript.canvas

import android.os.Handler
import android.os.Looper
import android.view.Choreographer
import android.view.Choreographer.FrameCallback
import java.util.*
import java.util.concurrent.Executors

/**
 * Created by triniwiz on 2019-08-13
 */
class AnimationFrame : FrameCallback {
	companion object {
		private var callbacks: HashMap<Long, (Long) -> Unit>? = null
		private var lastCall: Long = 0
		private var count = 0
		private var timer: Timer? = null
		private var instance: AnimationFrame? = null
		private var executorService = Executors.newCachedThreadPool()
		private val lock = Any()
		private var choreographer: Choreographer? = null
		private var handler: Handler? = null
		private var animationId: Long = 0
		private var _minFps = 1000
		private var framesRendered = 0
		private var frameStartTime: Long = 0
		private var inAnimationFrame = false
		private val timeInFrameBase: Long
			get() = System.nanoTime() / 1000000
		private val newId: Long
			get() {
				animationId++
				return animationId
			}

		@Synchronized
		fun requestAnimationFrame(callback: (Long) -> Unit): Long {
			if (!inAnimationFrame) {
				inAnimationFrame = true
				callback(timeInFrameBase)
				inAnimationFrame = false
				return newId
			}
			val id = newId
			callbacks!![id] = callback
			Choreographer.getInstance().postFrameCallback(instance)
			return id
		}

		fun cancelAnimationFrame(id: Long) {
			callbacks!!.remove(id)
		}

		init {
			callbacks = HashMap()
			instance = AnimationFrame()
			handler = Handler(Looper.getMainLooper())
		}
	}

	fun reset() {
		_minFps = 1000
		frameStartTime = 0
		framesRendered = 0
	}

	fun raf(fps: Long) {
		executorService.submit {
			val cbs = callbacks!!.clone() as HashMap<Long, (Long) -> Unit>
			callbacks!!.clear()
			inAnimationFrame = true
			val set: Set<Map.Entry<Long, (Long) -> Unit>> = cbs.entries
			for ((_, value) in set) {
				//if(cb.getKey() == animationId){
				handler!!.post { value(fps) }
				// }
			}
			inAnimationFrame = false
			reset()
		}
	}

	fun fps(currentTimeMillis: Long) {
		val fps: Int
		if (frameStartTime > 0) {
			// take the span in milliseconds
			val timeSpan = currentTimeMillis - frameStartTime
			framesRendered++
			if (timeSpan > 1000) {
				fps = (framesRendered * 1000 / timeSpan).toInt()
				if (fps < _minFps) {
					_minFps = fps
				}
				raf(fps.toLong())
				frameStartTime = currentTimeMillis
				framesRendered = 0
			}
		} else {
			frameStartTime = currentTimeMillis
		}
	}

	override fun doFrame(frameTimeNanos: Long) {
		fps(frameTimeNanos / 1000000)
		Choreographer.getInstance().postFrameCallback(instance)
	}
}
